### PR TITLE
Enum refactor: enum-based SELECT_CASE_var

### PR DESCRIPTION
**Pull request by dareumnam**
Pull request overview
---------------------
 -  Replace enum-based `SELECT_CASE_var` if-then ladders with switch/case statements.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by dareumnam** [comment]
Since this branch is getting bigger, I'd like to pick up comments on a follow-up PR. This is ready for review.

File: src/EnergyPlus/DataGlobalConstants.cc [hunk 1]
```diff
@@ -225,150 +225,148 @@ std::string GetResourceTypeChar(ResourceType const ResourceTypeNum)
...
-        } else if (SELECT_CASE_var == ResourceType::PlantLoopHeatingDemand) {
-            return "PlantLoopHeatingDemand";
-
-        } else if (SELECT_CASE_var == ResourceType::PlantLoopCoolingDemand) {
-            return "PlantLoopCoolingDemand";
-
-        } else {
-            return "Unknown";
-        }
+    switch (ResourceTypeNum) {
```
Comment (amirroth) [inline src/EnergyPlus/DataGlobalConstants.cc:228]: This should not be a function at all, it should be a `constexpr std::array<std::string_view>`.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 2]
```diff
@@ -1243,23 +1243,28 @@ std::string DisplayMaterialRoughness(DataSurfaces::SurfaceRoughness const Roughn
...
-        } else if (SELECT_CASE_var == DataSurfaces::SurfaceRoughness::MediumSmooth) {
-            cRoughness = "MediumSmooth";
-        } else if (SELECT_CASE_var == DataSurfaces::SurfaceRoughness::Smooth) {
-            cRoughness = "Smooth";
-        } else if (SELECT_CASE_var == DataSurfaces::SurfaceRoughness::VerySmooth) {
-            cRoughness = "VerySmooth";
-        } else {
-            cRoughness = "";
-        }
+    switch (Roughness) {
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc:1246]: Same comment.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 3]
```diff
@@ -1313,20 +1318,20 @@ Real64 ComputeNominalUwithConvCoeffs(EnergyPlusData &state,
...
-                    } else if (SELECT_CASE_var1 ==
-                               SurfaceClass::Floor) { // Interior:  horizontal, still air, heat flow downward, Rcin = 0.92 ft2-F-hr/BTU
-                        outsideFilm = 0.1620212;
-                    } else if (SELECT_CASE_var1 ==
-                               SurfaceClass::Roof) { // Interior:  horizontal, still air, heat flow upward, Rcin = 0.61 ft2-F-hr/BTU
-                        outsideFilm = 0.1074271;
-                    } else {
-                        outsideFilm = 0.0810106; // All semi-exterior surfaces
-                    }
+                switch (state.dataSurface->Surface(state.dataSurface->Surface(numSurf).ExtBoundCond).Class) {
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc:1321]: This can also be a constexpr `std::array<Real64>`.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 4]
```diff
@@ -1335,19 +1340,21 @@ Real64 ComputeNominalUwithConvCoeffs(EnergyPlusData &state,
...
-                insideFilm = 0.1197548;
-            } else if (SELECT_CASE_var == SurfaceClass::Floor) { // Interior:  horizontal, still air, heat flow downward, Rcin = 0.92 ft2-F-hr/BTU
-                insideFilm = 0.1620212;
-            } else if (SELECT_CASE_var == SurfaceClass::Roof) { // Interior:  horizontal, still air, heat flow upward, Rcin = 0.61 ft2-F-hr/BTU
-                insideFilm = 0.1074271;
-            } else {
-                insideFilm = 0.0;
-                outsideFilm = 0.0;
-            }
+        switch (state.dataSurface->Surface(numSurf).Class) {
```
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc:1343]: Ha!

File: src/EnergyPlus/DataSurfaces.cc [hunk 5]
```diff
@@ -666,47 +666,46 @@ std::string cSurfaceClass(SurfaceClass const ClassNo)
...
-        } else if (SELECT_CASE_var == SurfaceClass::Detached_B) {
-            ClassName = "Detached Shading:Building";
-
-        } else if (SELECT_CASE_var == SurfaceClass::Detached_F) {
-            ClassName = "Detached Shading:Fixed";
-
-        } else {
-            ClassName = "Invalid/Unknown";
-        }
+    switch (ClassNo) {
```
Comment (amirroth) [inline src/EnergyPlus/DataSurfaces.cc:669]: Same.

File: src/EnergyPlus/DesiccantDehumidifiers.cc [hunk 6]
```diff
@@ -2361,49 +2347,46 @@ namespace DesiccantDehumidifiers {
...
-
-                } else if (SELECT_CASE_var == PerformanceModel::UserCurves) {
-
-                    MinProcAirOutHumRat = CurveValue(state, DesicDehum(DesicDehumNum).ProcHumRatCurvefTW, ProcAirInTemp, ProcAirInHumRat) *
-                                          CurveValue(state, DesicDehum(DesicDehumNum).ProcHumRatCurvefV, ProcAirVel);
-
-                } else {
+            switch (DesicDehum(DesicDehumNum).PerformanceModel_Num) { // Performance Model Part A
+            case PerformanceModel::Default: {
+                WC0 = 0.0148880824323806;
```
Comment (amirroth) [inline src/EnergyPlus/DesiccantDehumidifiers.cc:2352]: `constexpr`?

File: src/EnergyPlus/DesiccantDehumidifiers.cc [hunk 7]
```diff
@@ -2419,110 +2402,109 @@ namespace DesiccantDehumidifiers {
...
+                ProcAirOutTemp = TC0 + TC1 * ProcAirInTemp + TC2 * ProcAirInHumRat + TC3 * ProcAirVel + TC4 * ProcAirInTemp * ProcAirInHumRat +
+                                 TC5 * ProcAirInTemp * ProcAirVel + TC6 * ProcAirInHumRat * ProcAirVel + TC7 * ProcAirInTemp * ProcAirInTemp +
+                                 TC8 * ProcAirInHumRat * ProcAirInHumRat + TC9 * ProcAirVel * ProcAirVel +
+                                 TC10 * ProcAirInTemp * ProcAirInTemp * ProcAirInHumRat * ProcAirInHumRat +
+                                 TC11 * ProcAirInTemp * ProcAirInTemp * ProcAirVel * ProcAirVel +
+                                 TC12 * ProcAirInHumRat * ProcAirInHumRat * ProcAirVel * ProcAirVel + TC13 * std::log(ProcAirInTemp) +
+                                 TC14 * std::log(ProcAirInHumRat) + TC15 * std::log(ProcAirVel);
+
+                // Regen energy
+                QC0 = -27794046.6291107;
```
Comment (amirroth) [inline src/EnergyPlus/DesiccantDehumidifiers.cc:2434]: `constexpr`?

File: src/EnergyPlus/EconomicTariff.cc [hunk 8]
```diff
@@ -4407,25 +4407,33 @@ void WriteTabularTariffReports(EnergyPlusData &state)
...
-                    } else if (SELECT_CASE_var == EconConv::MJ) {
-                        tableBody(1, 7) = "MJ";
-                    } else if (SELECT_CASE_var == EconConv::KBTU) {
-                        tableBody(1, 7) = "kBtu";
-                    } else if (SELECT_CASE_var == EconConv::MCF) {
-                        tableBody(1, 7) = "MCF";
-                    } else if (SELECT_CASE_var == EconConv::CCF) {
-                        tableBody(1, 7) = "CCF";
-                    }
+                switch (tariff(iTariff).convChoice) {
```
Comment (amirroth) [inline src/EnergyPlus/EconomicTariff.cc:4410]: `constexpr std::array<std::string_view>`

File: src/EnergyPlus/EvaporativeCoolers.cc [hunk 9]
```diff
@@ -1262,19 +1266,24 @@ void SizeEvapCooler(EnergyPlusData &state, int const EvapCoolNum)
...
-        } else if (SELECT_CASE_var == EvapCoolerType::IndirectWETCOIL) {
-            CompType = "EvaporativeCooler:Indirect:WetCoil";
-        } else if (SELECT_CASE_var == EvapCoolerType::IndirectRDDSpecial) {
-            CompType = "EvaporativeCooler:Indirect:ResearchSpecial";
-        } else if (SELECT_CASE_var == EvapCoolerType::DirectResearchSpecial) {
-            CompType = "EvaporativeCooler:Direct:ResearchSpecial";
-        } else if (SELECT_CASE_var == EvapCoolerType::DirectCELDEKPAD) {
-            CompType = "EvaporativeCooler:Direct:CelDekPad";
-        }
+    switch (EvapCond(EvapCoolNum).evapCoolerType) {
```
Comment (amirroth) [inline src/EnergyPlus/EvaporativeCoolers.cc:1269]: Same.

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 10]
```diff
@@ -4578,23 +4578,33 @@ void GetRoomAirModelParameters(EnergyPlusData &state, bool &errFlag) // True if
         {
             static constexpr std::string_view RoomAirZoneFmt("RoomAir Model,{},{}\n");
 
-            auto const SELECT_CASE_var(state.dataRoomAirMod->AirModel(ZoneNum).AirModelType);
-            if (SELECT_CASE_var == DataRoomAirModel::RoomAirModel::Mixing) {
+            switch (state.dataRoomAirMod->AirModel(ZoneNum).AirModelType) {
```
Comment (amirroth) [inline src/EnergyPlus/HeatBalanceAirManager.cc:4581]: `constexpr std::array<std::string_view>`

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 11]
```diff
@@ -684,208 +684,213 @@ void GatherForPredefinedReport(EnergyPlusData &state)
...
+                    ufactAreaNonNorth += nomUfact * windowAreaWMult;
+                    shgcAreaNonNorth += SHGCSummer * windowAreaWMult;
+                    vistranAreaNonNorth += TransVisNorm * windowAreaWMult;
+                }
+                // shading
+                if (Surface(iSurf).HasShadeControl) {
+                    PreDefTableEntry(state, state.dataOutRptPredefined->pdchFenSwitchable, surfName, "Yes");
+                    PreDefTableEntry(state, state.dataOutRptPredefined->pdchWscName, surfName, state.dataSurface->WindowShadingControl(curWSC).Name);
+                    // shading report
+                    std::vector<std::string> WindowShadingTypeStr = {
```
Comment (amirroth) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:817]: `constexpr std::array<std::string_view>` and same for the next one. May want to move to the .hh file also if they are replicated elsewhere.

File: src/EnergyPlus/PipeHeatTransfer.cc [hunk 12]
```diff
@@ -1209,23 +1214,27 @@ void PipeHTData::CalcPipesHeatTransfer(EnergyPlusData &state, Optional_int_const
...
-        } else {
-            EnvHeatTransCoef = 0.0;
-        }
+    switch (this->EnvironmentPtr) {
+    case EnvrnPtr::GroundEnv: {
+        // Approximate conductance using ground conductivity, (h=k/L), where L is grid spacing
+        // between pipe wall and next closest node.
+        EnvHeatTransCoef = this->SoilConductivity / (this->dSregular - (this->PipeID / 2.0));
+    } break;
+    case EnvrnPtr::OutsideAirEnv: {
```
Comment (amirroth) [inline src/EnergyPlus/PipeHeatTransfer.cc:1223]: same `case`?

**Comment by amirroth** [comment]
Looks good. You can leave my comments to a subsequent PR if you just want to get this thing merged. Thanks!

**Comment by dareumnam** [comment]
@amirroth Thank you for your comments. I'll pick up your comments on the next PR.

**Comment by dareumnam** [comment]
`EnergyPlusFixture.PackagedTerminalHP_VSCoils_Sizing` is passed on my local machine, but it fails in CI testing.

**Comment by Myoldmopar** [comment]
No problem @dareumnam, that one occasionally fails on CI, though I believe I found the fix for it in my DataHVACGlobals enum refactor that should be redone properly next week.  And those integration timeouts have popped up a few times lately; I'm going to give CI a reboot soon to see if I can clean that up.  I don't see where Windows worked on the last commit, though it shows up in the list of checks.  Hmm.  Well anyway, I don't think any of these oddities are related to your changes here, so I'm going to push forward on this branch review now.

**Review by Myoldmopar — APPROVED** [review]
Yep.  Yep, yep, yep.  This is such a fantastic cleanup.  Great work @dareumnam !

**Comment by Myoldmopar** [comment]
Passed with flying colors here, including regressions.  Merging this.  Thanks @dareumnam !

**Comment by amirroth** [comment]
Thanks @dareumnam, this PR reduced the number of SELECT_CASE_var statements from 3638 to 2481!

**Pull request merged**